feature: Add secrets troubleshooting [CY-1336] (#387)
codacy/chart: docs/configuration/secrets/troubleshoot.md

# Troubleshooting

The sections below include information to help you troubleshoot issues that you may come across while configuring the secrets used by Codacy.

If the information provided on this page is not enough to solve your issue, contact [support@codacy.com](mailto:support@codacy.com) providing all the information that you were able to collect while following the troubleshooting instructions.

## Lost/Changed initially generated secrets

While trying to open Codacy,
you were shown an error message stating that,
the secret on the database and the one in your configuration file are different.

### Steps

> 1. Get the tail of the logs on the api service

Reviewer: Three comments:

- Why mention tail specifically? I think "Get the latest logs from the api service" would be more appropriate. 
- Not sure if we are using "api" or "API".
- Guarantee that we are always calling 'api service' to the website backend.

Author: > Guarantee that we are always calling 'api service' to the website backend

Not sure how to do this.

The only reason I called it api is because the name of the pod is api.

> 1. Get the tail of the logs on the api service

Reviewer: Three questions:

- Why mention tail specifically? I think "Get the latest logs from the api service" would be more appropriate.
- Not sure if we are using 'api' or 'API'.
- Guarantee that we always use 'api service'  when referring to website's backend.

    a. Setup access to the cluster

    b. Get logs and filter by the prefix of the message

        ```sh

>         kubectl -n codacy logs deployment/codacy-api | grep -i "Your database is using key"

Author: Should we do a script similar to the one for the logs for this?
It has the same cluster setup reqs.
@afsmeira @jllopes

Reviewer 2: Yeah, I think it should, since it makes the troubleshooting easier

Reviewer: Yes, I agree.

Reviewer: Yes, I think it will help.

        ```

>         ```

Reviewer 2: Extra closing ``` here

> 2. Copy the value and update it in your values yaml file

Reviewer: Should we mention the specific config property to change?

Author: ths logs has it

3. Restart the application with the new configuration

    ```bash

>     helm upgrade (...options used to install Codacy...) \

Author: This line is starting to be repeated in too many places. I am starting to worry it will get outdated and cause issues.

Reviewer 2: Maybe we should have a document for commands repeated in multiple places and just start linking into that document? @pauloribeiro-codacy

Reviewer: That is not great @jllopes . Ideally we would have it in a script, but the script would still need to have the options mentioned above as input.

Reviewer 3: The `(...options used to install Codacy...)` bit was already meant to solve the issue of the command becoming outdated. Do you think that even the `helm upgrade` is subject to change?

Author: I do not think it will frequently.

# Troubleshooting

The sections below include information to help you troubleshoot issues that you may come across while configuring the secrets used by Codacy.

> If the information provided on this page is not enough to solve your issue, contact [support@codacy.com](mailto:support@codacy.com) providing all the information that you were able to collect while following the troubleshooting instructions.

Reviewer 3: This is a more generic comment, but I'm thinking we should check with Support if they're comfortable whenever we suggest to the users that they should use support@codacy.com to contact Codacy. At the very least, Support should know about these "entry points" so they are empowered to tackle the tickets from customers more efficiently. (@IGFCoimbra, @paulacodacy).

Author: As we prefer, I have no strong opinion about this.
I was just putting this were it makes sense for our customers to know where to go for help.

## Lost/Changed initially generated secrets

> While trying to open Codacy,
> you were shown an error message stating that,
> the secret on the database and the one in your configuration file are different.

Reviewer 3: ```suggestion
When you open the Codacy UI, an error message states that the secret on the database and the one in your configuration file are different.
```

### Steps

> 1. Get the latest logs from the API service

Reviewer 3: Since the script automatically extracts the value from the logs, we can simplify this a bit:

```suggestion
1. Obtain the correct key used to encrypt sensitive data in the database from the Codacy logs.
```

>     a. Get logs and filter by the prefix of the message

Reviewer 3: I would remove this detail (since the script does a great job in abstracting this) and unindent the lines below one level.

        ```bash
        bash <(curl -fsSL https://raw.githubusercontent.com/codacy/chart/master/docs/configuration/secrets/extract-codacy-secrets.sh) \
            -n <namespace>
        ```

        You can also download the script [extract-codacy-secrets.sh](extract-codacy-secrets.sh) to run it manually.

> 2. Copy the value and update it in your values yaml file

Reviewer 3: We can be just a bit more specific:

```suggestion
2. Copy the value of the key and update your `values-production.yaml` file with this value.
```

> 3. Restart the application with the new configuration

Reviewer 3: For consistency we can use exactly the same text here as we are using in the multiple existing places in the documentation (see step 4 of [this page](https://codacy.github.io/chart/configuration/integrations/github-cloud/), for instance).

If later we decide to extract and centralize this step somewhere else, it makes it easier to do the refactor if we already know that the instruction is exactly the same everywhere.
